### ISSUE TITLE
a problem occurs when running KD experiments

### ISSUE THREAD

**Issue by tdlhl** [Adamdad/KnowledgeFactor#5]
Excuse me, I wanted to follow this wonderful work. I have successfully conducted KF experiments on cifar10 and 
imagenet; however, a problem occurs when running KD experiments.
on cifar10, I run with: python tools/dist_train.py configs/cifar10-kd/wideresnet28-10_resnet18_b128x1
_cifar10.py 3
I get an error:
![image](https://user-images.githubusercontent.com/81509444/198066580-b31963d9-34f9-44c2-a4da-261f08e0a018.png)
so I changed line 159 of the /KnowledgeFactor-main/cls/mmcls/models/classifiers/kd.py with:
loss_cls = self.criterionCls(student_logit, gt_label.squeeze(dim=1))
However, training with the modified code achieved more than 96% acc,which is higher than the acc reported in the paper
![image](https://user-images.githubusercontent.com/81509444/198066719-c51c73bb-e851-44de-b415-23905e24d6c1.png)
So I think my modification is wrong, would you please tell me the right way? thank you very much!

**Comment by Adamdad** [comment]
Dear [tdlhl](https://github.com/tdlhl), What is your output and gt tensor size? For pytorch, nn.CrossEntropy(output, gt) should produce no error, where output is [batch_size x number_class], and gt is a [batch_size] tensor contains class index for each sample.

**Comment by tdlhl** [comment]
output(student_logits) [128,3,32,32]  gt[128,1]，so I use  ".squeeze(dim=1)" to change gt into [128]

**Comment by Adamdad** [comment]
Why your logins are a dense feature (Batch x Channel x H x W)? Did you change anything?

**Comment by tdlhl** [comment]
No, I didn't
![image](https://user-images.githubusercontent.com/81509444/198191575-fd529134-5b55-4cde-9a78-12890e68e596.png)

**Comment by Adamdad** [comment]
![image](https://user-images.githubusercontent.com/26020510/198192060-f02f2b8c-53a4-4f88-9450-e96bd9762cf9.png)
So you mean `student_logit` contains the spatial dimension? I dont think that is possible since I have pass it through the linear layer in the `self.get_logits` function.

**Comment by tdlhl** [comment]
Sorry, I was wrong. The shape of student_logits is [128, 10]

**Comment by Adamdad** [comment]
I personally think that is a torch version problem. In some torch versions, nn.functional.cross_entropy only allow (N) inputs, but in some version (N,) is allowed. And that should not effect the results. But I am still not sure why your results is pretty high.

https://pytorch.org/docs/stable/generated/torch.nn.functional.cross_entropy.html#torch.nn.functional.cross_entropy

**Comment by tdlhl** [comment]
my torch version is 1.12.1 
![image](https://user-images.githubusercontent.com/81509444/198193994-d91c3226-bd99-40cf-ae69-92cdca8fc7ff.png)
may I ask your torch version? so that I can change mine to run again
thank you very much

**Comment by Adamdad** [comment]
<img width="466" alt="image" src="https://user-images.githubusercontent.com/26020510/198195360-0de31139-4809-4244-93a7-52269232ecfe.png">
Mine is much lower, torch 1.6.0

**Comment by tdlhl** [comment]
OK, I'll try it later, thanks a lot

**Comment by Adamdad** [comment]
Please bring any questions back to me. Thanks. 

Also, there are some other standard distillation packages. This KF repo is with very simple KD use case, which may not be very extendable for distillation task itself. I will try to improve the repo.

https://github.com/haitongli/knowledge-distillation-pytorch
https://github.com/AberHu/Knowledge-Distillation-Zoo
https://github.com/HobbitLong/RepDistiller

**Comment by WangzcBruce** [comment]
> may I ask your torch version? so that I can change mine to run again
> thank you very much

Hello, have you reproduced the reported results？ I adopted your mentioned squeeze.dim solution to run the code and also found the same problem.  
![image](https://github.com/Adamdad/KnowledgeFactor/assets/45158420/471f0952-6ab4-47c9-89b9-348b733d5543)
![image](https://github.com/Adamdad/KnowledgeFactor/assets/45158420/109711ce-e19a-417e-849d-f75524593ad9)
